[inductor][quant]Enable inductor vec codegen for quantization

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98826
* #98730
* #99132
* #99131
* __->__ #98489

**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor &&  python -m pytest test_cpu_repro.py -k test_decomposed_dequant_relu_quant
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

leslie-fang-intel committed:
Enable inductor vec codegen for quantization

[ghstack-poisoned]

vadimkantorov commented on aten/src/ATen/cpu/vec/vec256/vec256_float.h:
```diff
@@ -97,6 +97,29 @@ template <> class Vectorized<float> {
       std::memcpy(ptr, tmp_values, count * sizeof(float));
     }
   }
+  void solution2_satuate(uint8_t* dst_data) {
```
typo? should be `saturate`?

leslie-fang-intel replied:
Oh, yean. Thanks. This PR is still work in progress. I will refine it sooner.

leslie-fang-intel replied:
Thanks, the typo is fixed.

leslie-fang-intel committed:
Update on "[WIP] Enable inductor vec codegen for quantization"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] Enable inductor vec codegen for quantization"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] Enable inductor vec codegen for quantization"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented on aten/src/ATen/cpu/vec/vec512/vec512_float.h:
```diff
@@ -121,6 +121,33 @@ template <> class Vectorized<float> {
       std::memcpy(ptr, tmp_values, count * sizeof(float));
     }
   }
+  void solution2_satuate(uint8_t* dst_data) {
```
Yes, thanks for the comment. The typo is fixed.

leslie-fang-intel committed:
Update on "Enable inductor vec codegen for quantization"


**Summary**
Enable the decomposed dequant - pointwise ops - decomposed quant, vectorization code gen inside inductor.
Take the example of `decomposed dequant - relu - decomposed quant` pattern as example, the generated code is as 
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor && python -m pytest test_torchinductor.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/cpu/vec/vec512/vec512_float.h:
```diff
@@ -121,6 +121,29 @@ template <> class Vectorized<float> {
       std::memcpy(ptr, tmp_values, count * sizeof(float));
     }
   }
+  void store_float_as_uint8(uint8_t* dst_data) {
```
It might be safe to do saturation and clamp as a general util function but seems not an efficient choice for inductor codegen if the code pattern is most likely quantization which has saturation and clamp before this particular intrinsics is called, therefore the logic here seems duplicate. It's probably ok to start with this slow path but can you explore the fast path too? Maybe we can match the pattern of quantization op in the loop body and invoke the fast path without explicit saturation and clamp?

leslie-fang-intel replied:
Thanks for the comment. Agree and I have marked it as a TODO and will optimize it in the following up PR after some init study of the new frontend API.

leslie-fang-intel committed:
Update on "Enable inductor vec codegen for quantization"


**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor && python -m pytest test_torchinductor.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented on test/inductor/test_torchinductor.py:
```diff
@@ -5926,6 +5926,51 @@ def fn(x):
+            not codecache.valid_vec_isa_list(), "Does not support vectorization"
+        )
+        @patch("torch.cuda.is_available", lambda: False)
+        def test_decomposed_dequant_relu_quant(self):
```
Done.

leslie-fang-intel committed:
Update on "[inductor][quant]Enable inductor vec codegen for quantization"


**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor &&  python -m pytest test_cpu_repro.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/cpu/vec/vec256/vec256_int.h:
```diff
@@ -1522,6 +1522,41 @@ Vectorized<uint8_t> inline operator>>(const Vectorized<uint8_t>& a, const Vector
+
+inline Vectorized<float> load_uint8_as_float(const uint8_t* src_data) {
+  // Load 8*uint8
+  __m256i input_256 = Vectorized<uint8_t>::loadu(src_data, 8);
```
The `loadu` with element number introduces additional temporary buffer. It might be better to use for-loop directly. Please check the performance.

leslie-fang-intel replied:
Thanks for the comment and changed. There is a intrinsic can implement it without using the `loadu`  of element numbers.

EikanWang commented on aten/src/ATen/cpu/vec/vec256/vec256_qint.h:
```diff
@@ -57,27 +57,20 @@ struct Vectorizedqi {
-inline __m256i pack_saturate_and_clamp<int32_t>(
-    __m256i /*first*/,
-    __m256i /*second*/,
+inline Vectorized<int32_t> pack_saturate_and_clamp<int32_t>(
+    Vectorized<int32_t> /*first*/,
+    Vectorized<int32_t> /*second*/,
```
Is this change neccessary for this PR?

leslie-fang-intel replied:
Thanks and changed.

jgong5 commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -1161,4 +1071,36 @@ inline void transpose_mxn(const T* src, int64_t ld_src, T* dst, int64_t ld_dst)
   }
 }
 
+#if (defined(CPU_CAPABILITY_AVX512) || defined(CPU_CAPABILITY_AVX2)) && !defined(_MSC_VER)
```
This is necessary?

leslie-fang-intel replied:
No, it's not necessary. I prefer to add it for the function declaration because the corresponding implementation defined with same macro. I thin we can remove it also.

leslie-fang-intel replied:
I have removed it.

leslie-fang-intel committed:
Update on "[inductor][quant]Enable inductor vec codegen for quantization"


**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor &&  python -m pytest test_cpu_repro.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][quant]Enable inductor vec codegen for quantization"


**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor &&  python -m pytest test_cpu_repro.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@jansel @jerryzh168 Could you help to take a look of this PR to enable the decomposed quant/dequant vectorization code gen.

leslie-fang-intel committed:
Update on "[inductor][quant]Enable inductor vec codegen for quantization"


**Summary**
Enable the `decomposed dequant - pointwise ops - decomposed quant` vectorization code gen inside inductor.
Here is the example in the UT and the generated code:

Example:
* `decomposed dequant - relu - decomposed quant` pattern.
* Using `uint8` as the quantized tensor data type.

Generated Code:
```
kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/hw/chwr6vy6e6sd25sfh42qtywkuf2emodexm2aomp3lbrcxwznfwyi.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       unsigned char* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0); i0<static_cast<long>(27); i0+=static_cast<long>(1))
            {
                auto tmp0 = at::vec::load_uint8_as_float(in_ptr0 + static_cast<long>(16*i0));
                auto tmp1 = (tmp0);
                auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(100));
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(0.01));
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = at::vec::clamp_min(tmp5, decltype(tmp5)(0));
                auto tmp7 = at::vec::Vectorized<float>(static_cast<float>(100.0));
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = tmp8.round();
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = at::vec::Vectorized<float>(static_cast<float>(0));
                auto tmp12 = at::vec::maximum(tmp10, tmp11);
                auto tmp13 = at::vec::Vectorized<float>(static_cast<float>(255));
                auto tmp14 = at::vec::minimum(tmp12, tmp13);
                auto tmp15 = (tmp14);
                tmp15.store_float_as_uint8(out_ptr0 + static_cast<long>(16*i0));
            }
            #pragma omp for simd simdlen(8)
            for(long i0=static_cast<long>(432); i0<static_cast<long>(441); i0+=static_cast<long>(1))
            {
                auto tmp0 = in_ptr0[static_cast<long>(i0)];
                auto tmp1 = static_cast<float>(tmp0);
                auto tmp2 = static_cast<float>(100);
                auto tmp3 = tmp1 - tmp2;
                auto tmp4 = static_cast<float>(0.01);
                auto tmp5 = tmp3 * tmp4;
                auto tmp6 = tmp5 * (tmp5>0);
                auto tmp7 = static_cast<float>(100.0);
                auto tmp8 = tmp6 * tmp7;
                auto tmp9 = std::nearbyint(tmp8);
                auto tmp10 = tmp9 + tmp2;
                auto tmp11 = static_cast<float>(0);
                auto tmp12 = (tmp11 != tmp11) ? tmp11 : std::max(tmp10, tmp11);
                auto tmp13 = static_cast<float>(255);
                auto tmp14 = (tmp13 != tmp13) ? tmp13 : std::min(tmp12, tmp13);
                auto tmp15 = static_cast<unsigned char>(tmp14);
                out_ptr0[static_cast<long>(i0)] = tmp15;
            }
        }
    }
}
''')
```

**Test Plan**
```
cd test/inductor &&  python -m pytest test_cpu_repro.py -k test_decomposed_dequant_relu_quant
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
